Day029-router-lazy-load.md

```diff
@@ -0,0 +1,277 @@
+   Note: nhớ xóa cả phần `import` ở trên đi nhá, không thì sẽ ko giải quyết vấn đề gì đâu :))
+
+![Step 6][step06]
+
+2. Config lại route cho `AdminModule`. Mình cần bỏ phần path admin ở trong `AdminRoutingModule` đi. Lý do thì các bạn xem step 3 sẽ rõ.
+
+![Step 7][step07]
+
+3. Bước quan trọng nhất, trong `AppRoutingModule`, chúng ta cần config để load admin module theo cú pháp của lazy loading module khi mở ứng dụng ở path có dạng `/admin`
+
+![Step 8][step08]
+
+Vẫn là định nghĩa là path `admin`, nhưng thay vì component, mình dùng `loadChildren` với cú pháp `() => import('./admin/admin.module').then((m) => m.AdminModule)`, tức là một function có return lại một dynamic import module.
+
+> Notice that the lazy-loading syntax uses loadChildren followed by a function that uses the browser's built-in import('...') syntax for dynamic imports. The import path is the relative path to the module.
+
+Vậy là xong rồi đấy. Giờ mình mở sẽ run `npm run build` một lần nữa. Giờ CLI đã tạo ra một file JS mới dành riêng cho `AdminModule` tên `admin-admin-module.js`.
+
+![Step 9][step09]
+
+Mình sẽ mở app lên xem thử, các bạn sẽ thấy một sự khác bọt khá lớn.
+
+![Step 10][step10]
+
+Khi mình mới load page, thì app có load các file cần thiết và `main.js`. Tuy nhiên khi mình ấn vào Admin để đi đến `/admin`, lúc đó `admin-admin-module.js` mới được load. Lazy load chính là như thế, tức là không load ngay từ đầu, mà chỉ load khi cần đến.
+
+Giờ chúng ta cùng xem bundle của `main.js` và `admin.module.js` nhé.
+
+```
+source-map-explorer main.js
+source-map-explorer admin-admin-module.js
+```
+
+![Step 11][step11]
+
+Phần bundle của `main.js` giờ đã không còn AdminModule nữa rồi, chỉ còn ArticleModule thôi.
+
+![Step 12][step12]
+
+Còn đây là phân tích của `admin-admin-module.js`. Phần import của Form và các component của AdminModule đã nằm trong bundle này. Quá xuất sắc.
+
+### Preloading Lazy Module
+
+Việc tách ra thành các lazy loading module rất có lợi cho lần tải trang của user. Với việc browser phải download file JS nhỏ hơn, nên thời gian chờ đợi để tương tác với website sẽ giảm. Tuy nhiên nhiều khi các lazy loading module có thể có size rất lớn. Dẫn đến việc khi user bấm vào một link, sẽ mất thời gian để tải. Ví dụ vẫn là lazy loading `AdminModule` ở trên nhưng giờ mình giả lập là mạng điện thoại, xem tình hình thế nào nhé.
+
+![Step 13][step13]
+
+Anh em thấy đấy, thời gian từ khi bấm vào link Admin cho đến khi UI hiện ra có khi cả 10s. Quá chậm.
+
+Có một số module mà mình biết rằng thường là khi user mở ứng dụng của mình ra, họ sẽ click vào đó đầu tiên. Chúng ta có thể dùng kĩ thuật Preloading để load những bundle ở background.
+
+Để enable preloading cho tất cả các lazy loaded modules, các bạn cần import `PreloadAllModules` từ package `@angular/router` và cấu hình nó ở trong AppRoutingModule, đoạn forRoot.
+
+```ts
+import { PreloadAllModules } from "@angular/router";
+
+const routes: Routes = [
+  {
+    path: "admin",
+    loadChildren: () =>
```

Comment:
Just FYI, don't need to modify anything (unless you want to)

`loadChildren: async() => (await import('./admin/admin.module')).AdminModule`
or

```ts
loadChildren: () => from(import('./admin/admin.module')).pipe(
   map(m => m.AdminModule),
   retry(2) // retry twice :3
)
```